src/i18n/util.ts

```diff
@@ -1,9 +1,28 @@
+import type { AstroGlobal } from 'astro';
+import { getLanguageFromURL } from '../util';
 import { translations } from './translations';
 
 const fallbackLang = 'en';
 
 type Keys = keyof typeof translations[typeof fallbackLang];
 
-export function getLanguageString(key: Keys, lang = 'en'): string | undefined {
+function getLanguageString(key: Keys, lang = 'en'): string | undefined {
 	return translations[lang]?.[key] || translations[fallbackLang][key];
 }
+
+/**
+ * Get a UI string to pass to a framework component.
+ * 
+ * Within an Astro component, prefer the `UIString` component,
+ * which only needs the key as it has access to the global Astro object.
+ * 
+ * However, you can’t pass an Astro component as a prop to a framework component,
+ * so this function is provided to get the string instead:
+ * 
+ * @example
+ * <FrameworkComponent label={getUIString('new', Astro)} />
+ */
+export function getUIString(key: Keys, Astro: Readonly<AstroGlobal>) {
```

Comment:
You know, I hadn’t even thought, but this function could theoretically be:

```js
function getUIString(key: Keys, Astro?: Readonly<AstroGlobal>) {
  const { pathname } = Astro ? Astro.request.canonicalURL : window.location;
  const lang = getLanguageFromURL(pathname);
  return getLanguageString(key, lang);
}
```

And just call it without the second argument in framework components. (Although can a framework component use `window` during build? I guess not, so this would be `client:only` only.)

One benefit of the approach in this PR is that the only extra data shipped to the browser is the prop needed to hydrate the component, e.g. `{"prop":"translation"}`. Whereas with a more universal approach like the above, you’d need to send the browser the full multilingual dictionary.

The Astro/non-Astro component boundary is for sure an interesting space!

Reply:
(Thanks for the great review.)

Reply:
Yea, the problem is that the `Astro` global only exists inside of Astro components :(